app/modules/asset.rb

```diff
@@ -0,0 +1,148 @@
+class Asset
+
+  DEFAULT_START_DATE = 8.days.ago.strftime('%F')
+  DEFAULT_END_DATE = 1.day.ago.strftime('%F')
+
+
+  attr_reader :basic_stats,
+              :sources,
+              :os_sources,
+              :traffic_sources,
+              :country_sources,
+              :service,
+              :params,
+              :property,
+              :current_start_date,
+              :current_end_date,
+              :previous_period,
+              :current_period
+
+
+  def initialize(service, params)
+    @service = service
+    @params = params
+    @property = get_property
+    @basic_stats = get_basic_stats
+    @sources = get_sources
+    @os_sources = get_os_sources
+    @traffic_sources = get_traffic_sources
+    @country_sources = get_country_sources
+    @previous_pariod, @current_period = basic_stats.rows.each_slice(basic_stats.rows.size/2).to_a
+    @current_start_date = set_date(current_period[0][0])
+    @current_end_date = set_date(current_period[-1][0])
+  end
+
+
+  def change_in_percent
+    "#{change}%"
+  end
+
+  def change_status
+    change.positive? ? 'increase' : 'decrease'
+  end
+
+  def current_visits
+    summarize_visits(@current_period)
+  end
+
+  def previous_visits
+    summarize_visits(@previous_pariod)
+  end
+
+  def current_page_views
+    summarize_page_views(@current_period)
+  end
+
+  def previous_page_views
+    summarize_page_views(@previous_pariod)
+  end
+
+  def image
+    ReportGenerator.generate(self)
```

Comment:
With this method, the `ReportGenerator` is strongly coupled to the report asset itself. Would it make sense to move this logic higher up into the controller and let the controller decide how the reports will be generated? It feels like this method is mixing two different concerns into the class (most of the class is concerned with the in-memory _representation_ and _collection_ of the data, whereas this method is more about _presentation_)